### DISCUSSION TITLE
[Help wanted] Integrate custom shell

### DISCUSSION THREAD

**Discussion by Anakael** [hecrj/iced#1214] [Q&A]
I'd like to use iced with `wlr_layer_shell` protocol, but neither `glutin` or `winit` is acceptable for that.
So, is it possible to implement custom shell to integrate in iced? If so, may I ask you some guidelines for that?

**Comment by dominik-muc** [comment]
@hecrj Any plans about this? It's an old request, but so far most viewed one. I would like to participate in implementing this, however I'd use some guidelines on how to approach it. And if it's even gonna get merged, or is there an issue with this functionality? I know folks from cosmic de used layer shell, so maybe we should consult them and try to merge branches? Cheers

**Comment by Gibbz** [comment]
wanting to to know to use the layer shell with iced also

**Comment by Koranir** [comment]
As of now, the best way to use layer shell with iced is to use the pop!_os fork, which has a custom wayland backend behind a feature flag that adds Wayland-specific settings including the  layer.

**Reply by dominik-muc** [reply]
> As of now, the best way to use layer shell with iced is to use the pop!_os fork, which has a custom wayland backend behind a feature flag that adds Wayland-specific settings including the layer.

Why not integrate with it?

**Reply by Koranir** [reply]
It is a *lot* of custom, Wayland-specific code that would need to be maintained alongside the existing winit backend, that also has a different api than the winit backend. It adds a lot of maintenance burden for features that the vast majority of users will never touch, so it's better to leave it to the popos team that do require it and are willing to work on it..

**Comment by Exidex** [comment]
There is also https://github.com/waycrate/exwlshelleventloop/tree/master/iced_layershell

**Comment by Hillard28** [comment]
Think the general flow of events (at least w.r.t. `winit` support) is: there are outstanding PRs adding `wlr_layer_shell` support to `winit`, but `winit` contributors had stated they didn't intend to go ahead with any until they finished splitting the various modules of `winit` (e.g. `winit-wayland`). I believe they recently completed this split, and at least one author has stated their intent to align their PR with the new `winit-wayland` module. Hopefully that means `wlr_layer_shell` support is coming soon<sup>TM</sup>, at least for `winit`! Basically checkout PRs [#4044](https://github.com/rust-windowing/winit/pull/4044) and [#3433](https://github.com/rust-windowing/winit/issues/3433).

**Comment by DamitusThyYeetus123** [comment]
Looks like the consensus from the winit guys is this should be integrated into iced, not winit, so I think we should start having a look at that. I might start work on a fork to get it going without relying on glutin or winit.

**Comment by d-itu** [comment]
I found this `iced::Program`, and i'm replacing the winit runner with a custom one based on `libwayland`.
```rust
    let program = iced::daemon(Example::new, Example::update, Example::view)
        .subscription(Example::subscription)
        .title(Example::title)
        .theme(Example::theme)
        .scale_factor(Example::scale_factor);

    Ok(iced_winit::run(program)?)

```
The code is from [examples/multi-window](https://github.com/iced-rs/iced/blob/master/examples/multi_window/src/main.rs), with something modified. What I'm doing is to replace the `iced_winit::run` to my own winit-free `run`.

However, I found a lot of things deeply bundled with winit, making it hard to rewrite a different one. For example `window::Action` and `window::Settings` do not contain what i need to implement a wlr event loop such as surface role and surface role specific things etc.

I think a possible solution is to abstract window things like `Action` and `Settings` into an associated type in `Program` trait. To do this, something in core::window need to be moved to winit, remaining only winit-free things.
Or at least add some customizable fields in `window::Settings`.

Once those abstractions are done, writing any event-loops (or maybe shell?) could be easy.

**Reply by d-itu** [reply]
Another way to combine iced with native wayland is like [examples/integration](https://github.com/iced-rs/iced/tree/master/examples/integration). This is also the only solution i can come up with without forking iced.
However this maybe requires writing more code to manage renderer and ui states (and easier to cause bug), comparing to the above one.
